plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/server/N4JSStatefulIncrementalBuilder.java

```diff
@@ -29,4 +141,173 @@ protected void unloadResource(URI uri) {
 			super.unloadResource(uri);
 		}
 	}
+
+	/**
+	 * This override introduces two changes:
+	 * <p>
+	 * (1) on initial/full builds, all uris are sorted.
+	 * <p>
+	 * (2) on closure builds, the build request is adjusted so that dirty/deleted files always respect the URI closure.
+	 */
+	@Override
+	protected XBuildRequest initializeBuildRequest(XBuildRequest initialRequest, XBuildContext context) {
+		ProjectBuilder projectBuilder = workspaceManager.getProjectBuilder(initialRequest.getProjectName());
+		N4JSProjectConfigSnapshot projectConfig = (N4JSProjectConfigSnapshot) projectBuilder.getProjectConfig();
+		ResourceDescriptionsData oldIndex = context.getOldIndex();
+		boolean isInitialBuild = oldIndex.isEmpty() && initialRequest.getDeletedFiles().isEmpty();
+
+		if (projectConfig.hasTsConfigBuildSemantic()) {
```

Comment:
Currently the "closure build mode" is turned on/off for an entire project. But wouldn't it make sense to decide this based on the file type: `.d.ts` files are always built according to "closure build mode"; `.n4js[d]` etc. files are built according to "N4JS build mode"? (even within a single project)

This would then cover cases in which .d.ts files are located in N4JS projects.

(Maybe, `.d.ts` files located in an N4JS source folder inside an N4JS project (i.e. with projectType != PLAINJS) could also be built according to "N4JS build mode".)

In any case, this could be refined in the future; not intended as a suggestion for this PR.